### PR TITLE
[FEATURE] implementing product prices for different currencies

### PR DESCRIPTION
**Pull request by michaelzangerle**
__Tasks:__
- [x] done
- [x] got feedback

__Informations:__

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Tests pass?   | yes
| Fixed tickets | https://github.com/massiveart/POOL-ALPIN/issues/859
| Doc           | none

### CODE REVIEW

File: Entity/SpecialPrice.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+{
+    /**
+     * @var float
+     */
+    private $price;
+
+    /**
+     * @var \DateTime
+     */
+    private $start;
```
Comment (turbo-ele) [inline Entity/SpecialPrice.php]: from

File: Entity/SpecialPrice.php [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * @var \DateTime
+     */
+    private $end;
```
Comment (turbo-ele) [inline Entity/SpecialPrice.php]: to

File: Entity/SpecialPrice.php (hunk 1 above)
Comment (turbo-ele) [inline Entity/SpecialPrice.php]: or 
startDate

File: Entity/SpecialPrice.php (hunk 2 above)
Comment (turbo-ele) [inline Entity/SpecialPrice.php]: or 
endDate

File: Product/ProductManager.php [hunk 3]
```diff
@@ -815,6 +815,12 @@ public function save(
             $product->setParent($parentProduct);
         }
 
+        if (isset($data['cost']) && is_float($data['cost'])) {
```
Comment (turbo-ele) [inline Product/ProductManager.php]: is_numeric

File: Product/ProductManager.php [hunk 4]
```diff
@@ -815,6 +815,12 @@ public function save(
...
+            if (is_float($data['cost'])) {
```
Comment (turbo-ele) [inline Product/ProductManager.php]: remove

File: Product/ProductManager.php [hunk 5]
```diff
@@ -1525,4 +1531,8 @@ public function getFilters(Request $request)
         }
         return $filter;
     }
+
+    public function getDefaultCurrency(){
```
Comment (turbo-ele) [inline Product/ProductManager.php]: doc

File: Resources/public/js/components/bulk-price/main.js [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        INITIALIZED = function() {
+            return createEventName.call(this, 'initialized');
+        },
+
+        /**
+         * Returns the sales price (price with minimum quantity 0) and formats prices according locale
+         * @param prices
+         * @returns price
+         */
+        getSalesPrice = function(prices) {
```
Comment (turbo-ele) [inline Resources/public/js/components/bulk-price/main.js]: function name is missleading

File: Resources/public/js/components/bulk-price/main.js [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        /**
+         * @event sulu.products.bulk-price.initialized
+         * @description Emitted when component is initialized
+         */
+        INITIALIZED = function() {
+            return createEventName.call(this, 'initialized');
+        },
+
+        /**
+         * Returns the sales price (price with minimum quantity 0) and formats prices according locale
```
Comment (turbo-ele) [inline Resources/public/js/components/bulk-price/main.js]: ... and splits salesprice from prices array

File: Resources/public/js/components/bulk-price/main.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                for (; i < constants.maxBulkElements; i++) {
+                    prices.push({});
+                }
+            }
+
+            return prices;
+        },
+
+        bindCustomEvents = function() {
+            this.sandbox.on('sulu.products.bulk-price.get-data', function() {
```
Comment (turbo-ele) [inline Resources/public/js/components/bulk-price/main.js]: remove

File: Resources/public/js/components/bulk-price/main.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+                salesPrice,
+
+                $prices = this.sandbox.dom.find('.table tbody tr', this.$el),
+                priceValue, priceQuantity, priceId;
+
+            // update sales price
+            if (!!salesPriceValue) {
+                salesPrice = {
+                    price: !!salesPriceValue ? this.sandbox.parseFloat(salesPriceValue) : null,
+                    minimumQuantitiy: 0,
```
Comment (turbo-ele) [inline Resources/public/js/components/bulk-price/main.js]: constant for minimumQuantity

File: Resources/public/js/components/bulk-price/main.js [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+         * Returns the sales price (price with minimum quantity 0) and formats prices according locale
+         * @param prices
+         * @returns price
+         */
+        getSalesPrice = function(prices) {
+            var salesPrice = null,
+                idx = null;
+
+            this.sandbox.util.foreach(prices, function(price, index) {
+                if (parseFloat(price.minimumQuantity) === 0 && idx === null) {
```
Comment (turbo-ele) [inline Resources/public/js/components/bulk-price/main.js]: use minimumQuantity constant

File: Resources/public/js/components/bulk-price/main.js [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+            // bulk prices
+            this.sandbox.util.foreach($prices, function($price) {
+                priceId = this.sandbox.dom.data($price, 'id');
+                priceQuantity = this.sandbox.dom.val(this.sandbox.dom.find('input.minimumQuantity',$price));
+                priceValue = this.sandbox.dom.val(this.sandbox.dom.find('input.price',$price));
+
+                if (!!priceQuantity && !!priceValue) {
+                    priceItems.push({
+                        minimumQuantity: !!priceQuantity ? this.sandbox.parseFloat(priceQuantity) : null,
```
Comment (turbo-ele) [inline Resources/public/js/components/bulk-price/main.js]: remove if

File: Resources/public/js/components/bulk-price/main.js [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+                        price: !!priceValue ? this.sandbox.parseFloat(priceValue) : null,
```
Comment (turbo-ele) [inline Resources/public/js/components/bulk-price/main.js]: remove if

File: Entity/SpecialPrice.php (hunk 1 above)
Comment (michaelzangerle) [inline Entity/SpecialPrice.php]: look here https://camo.githubusercontent.com/a32cdd182067c7195a0c80d5d474fdc4cd425645/68747470733a2f2f7261772e6769746875622e636f6d2f6d6173736976656172742f73756c752d646f63732f6d61737465722f64657461696c2d73706563696669636174696f6e2f696d616765732f64622f70726f64756374732e706e67

File: Entity/SpecialPrice.php (hunk 2 above)
Comment (michaelzangerle) [inline Entity/SpecialPrice.php]: look here https://github.com/sulu-cmf/docs/blob/master/detail-specification/600-global/DET-601-products.md

File: Resources/public/js/components/products/components/pricing/main.js [hunk 13]
```diff
@@ -38,14 +57,22 @@ define(['config'], function (Config) {
                 save.call(this);
             }, this);
 
-            this.sandbox.on('sulu.products.saved', function () {
+            this.sandbox.on('sulu.products.saved', function (data) {
```
Comment (turbo-ele) [inline Resources/public/js/components/products/components/pricing/main.js]: format?

File: Resources/translations/sulu/backend.en.xlf [hunk 14]
```diff
@@ -202,14 +202,30 @@
...
                 <target>Price (EUR)</target>
             </trans-unit>
-            <trans-unit id="53" resname="product.price.CHF">
+            <trans-unit id="56" resname="product.price.CHF">
                 <source>product.price.CHF</source>
                 <target>Price (CHF)</target>
             </trans-unit>
+            <trans-unit id="57" resname="product.bulkprice.andMore">
+                <source>product.bulkprice.andMore</source>
+                <target>abnd more ...</target>
```
Comment (turbo-ele) [inline Resources/translations/sulu/backend.en.xlf]: typo

File: Entity/SpecialPrice.php (hunk 1 above)
Comment (turbo-ele) [inline Entity/SpecialPrice.php]: doesn't mean i have to like it :P

**Comment by turbo-ele** [comment]
ack

**Pull request merged**